File: ThreadLab4C#/ThreadLab4/ThreadLab4WithWaiter/Philosopher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ThreadLab4WithWaiter;

namespace ThreadLab4
{
    public class Philosopher
    {
        private readonly Table _table;
        private readonly Waiter _waiter;
        private readonly int _leftFork;
        private readonly int _rightFork;
        private readonly int _id;

        public Philosopher(Table table, int id, Waiter waiter)
        {
            _table = table;
            _id = id;
            _leftFork = id;
            _rightFork = (id + 1) % 5;
            _waiter = waiter;
        }

        public void Run()
        {
            for (int i = 0; i < 10; i++)
            {
                Console.WriteLine("Philosopher " + _id + " is thinking " + (i + 1) + " times");
                _waiter.RequestPermission();
                _table.GetFork(_leftFork);
                _table.GetFork(_rightFork);
                Console.WriteLine("Philosopher " + _id + " is eating " + (i + 1) + " times");
                _table.PutFork(_leftFork);
                _table.PutFork(_rightFork);
                _waiter.ReleasePermission();
            }
        }

    }
}

File: ThreadLab4Java/ThreadLab4Waiter/src/Philosopher.java
public class Philosopher implements Runnable {
    private final Table table;
    private final Waiter waiter;
    private final int leftFork;
    private final int rightFork;
    private final int id;

    public Philosopher(Table table, int id, Waiter waiter) {
        this.table = table;
        this.id = id;
        this.leftFork = id;
        this.rightFork = (id + 1) % 5;
        this.waiter = waiter;
    }

    @Override
    public void run() {
        for (int i = 0; i < 10; i++) {
            System.out.println("Philosopher " + id + " is thinking " + (i + 1) + " times");
            waiter.requestPermission();
            table.getFork(leftFork);
            table.getFork(rightFork);
            System.out.println("Philosopher " + id + " is eating " + (i + 1) + " times");
            table.putFork(leftFork);
            table.putFork(rightFork);
            waiter.releasePermission();
        }
        
    }
}